src/lib/block_producer/block_producer.ml

```diff
@@ -810,66 +905,173 @@ let run ~logger ~prover ~verifier ~trust_system ~get_completed_work
-                  (let now = Time.now time_controller in
-                   let%bind next_producer_timing =
-                     measure "asking consensus what to do" (fun () ->
-                         Consensus.Hooks.next_producer_timing
-                           ~constraint_constants ~constants:consensus_constants
-                           (time_to_ms now) consensus_state
-                           ~local_state:consensus_local_state ~keypairs
-                           ~coinbase_receiver:!coinbase_receiver ~logger )
+                  (let%bind () =
+                     if Mina_numbers.Length.(new_epoch > epoch) then
+                       Vrf_evaluation_state.update_epoch_data ~vrf_evaluator
+                         ~epoch_data_for_vrf ~logger vrf_evaluation_state
+                     else Deferred.unit
+                   in
+                   let%bind () =
+                     (*Poll once every slot if the evaluation for the epoch is not completed or the evaluation is completed*)
+                     if
+                       Mina_numbers.Global_slot.(new_global_slot > slot)
+                       && not
+                            (Vrf_evaluation_state.finished vrf_evaluation_state)
+                     then
+                       Vrf_evaluation_state.poll ~vrf_evaluator ~logger
+                         vrf_evaluation_state
+                     else Deferred.unit
                    in
-                   set_next_producer_timing next_producer_timing
-                     consensus_state ;
-                   match next_producer_timing with
-                   | `Check_again time ->
-                       Singleton_scheduler.schedule scheduler (time_of_ms time)
-                         ~f:check_next_block_timing ;
-                       Deferred.return ()
-                   | `Produce_now (data, winner_pk) ->
-                       Mina_metrics.(Counter.inc_one Block_producer.slots_won) ;
-                       let%map () = generate_genesis_proof_if_needed () in
-                       Interruptible.finally
-                         (Singleton_supervisor.dispatch production_supervisor
-                            (now, data, winner_pk))
-                         ~f:check_next_block_timing
-                       |> ignore
-                   | `Produce (time, data, winner_pk) ->
-                       Mina_metrics.(Counter.inc_one Block_producer.slots_won) ;
-                       let scheduled_time = time_of_ms time in
-                       don't_wait_for
-                         ((* Attempt to generate a genesis proof in the slot
+                   match Core.Queue.dequeue vrf_evaluation_state.queue with
+                   | None ->
+                       (*Keep trying until we get some slots*)
+                       if
+                         not
+                           (Vrf_evaluation_state.finished vrf_evaluation_state)
+                       then
+                         let%bind () =
```

Comment:
Can we use a blocking `Vrf_evaluator.found_new_or_finished` RPC here, that returns only when the evaluator has found the next won slot, or is finished?